core/store/src/trie/mem/updating.rs

```diff
@@ -0,0 +1,1281 @@
+        node_id: UpdatedMemTrieNodeId,
+        // The current extension.
+        extension: Box<[u8]>,
+        // The current child.
+        child_id: OldOrUpdatedNodeId,
+    ) {
+        let child_id = self.ensure_updated(child_id);
+        let child_node = self.take_node(child_id);
+        match child_node {
+            UpdatedMemTrieNode::Empty => {
+                // This case is not possible. In a trie in general, an extension
+                // node could only have a child that is a branch (possibly with
+                // value) node. But a branch node either has a value and at least
+                // one child, or has at least two children. In either case, it's
+                // impossible for a single deletion to cause the child to become
+                // empty.
+                unreachable!();
+            }
+            // If the child is a leaf (which could happen if a branch node lost
+            // all its branches and only had a value left, or is left with only
+            // one branch and that was squashed to a leaf).
+            UpdatedMemTrieNode::Leaf { extension: child_extension, value } => {
+                let child_extension = NibbleSlice::from_encoded(&child_extension).0;
+                let extension = NibbleSlice::from_encoded(&extension)
+                    .0
+                    .merge_encoded(&child_extension, true)
+                    .into_vec()
+                    .into_boxed_slice();
+                self.place_node(node_id, UpdatedMemTrieNode::Leaf { extension, value })
+            }
+            // If the child is a branch, there's nothing to squash.
+            child_node @ UpdatedMemTrieNode::Branch { .. } => {
+                self.place_node(child_id, child_node);
+                self.place_node(
+                    node_id,
+                    UpdatedMemTrieNode::Extension {
+                        extension,
+                        child: OldOrUpdatedNodeId::Updated(child_id),
+                    },
+                );
+            }
+            // If the child is an extension (which could happen if a branch node
+            // is left with only one branch), join the two extensions into one.
+            UpdatedMemTrieNode::Extension { extension: child_extension, child: inner_child } => {
+                let child_extension = NibbleSlice::from_encoded(&child_extension).0;
+                let merged_extension = NibbleSlice::from_encoded(&extension)
+                    .0
+                    .merge_encoded(&child_extension, false)
+                    .into_vec()
+                    .into_boxed_slice();
+                self.place_node(
+                    node_id,
+                    UpdatedMemTrieNode::Extension {
+                        extension: merged_extension,
+                        child: inner_child,
+                    },
+                );
+            }
+        }
+    }
+
+    fn post_order_traverse_updated_nodes(
```

Comment:
Comment with smth like "Needed to ensure that hashes of nodes can be computed from left to right in resulting list"